### PR TITLE
d/aws_vpc_endpoint_service - filter support

### LINKED ISSUE #11168: aws_vpc_endpoint_service requires additional filter block or name resolution
_This issue was originally opened by @TrickMcCarthy as hashicorp/terraform#23591. It was migrated here as a result of the [provider split](https://www.hashicorp.com/blog/upcoming-provider-changes-in-terraform-0-10/). The original body of the issue is below._
    
<hr>
    
### SDK version

```
aws provider 2.40
```

### Use-cases

https://www.terraform.io/docs/providers/aws/d/vpc_endpoint_service.html
This only supports service_name and service. It is missing support for the Name attribute.
The service_name and service are auto-generated each time the service endpoints are created.
The only way to utilize the data source is to hardcode the generated service_name or service id.
In the case I am experiencing the service endpoint is generated by a different service with different terraform state file.

Without supporting retrieval using an end user specified name, this hardcoded value needs to be manually updated each time the service endpoint is recreated. This creates a tight coupling between the services and requires coordination whenever the service endpoint is recreated. It would be much better to reduce this tight coupling.

### Attempted Solutions

While the remote_state can be used as a workaround, but it leaves things with tight coupling and this coupling is not suitable when you are trying to upgrade terraform between versions where the state file formats change.
https://www.terraform.io/docs/providers/terraform/d/remote_state.html

### Proposal

Add filter block support or retrieval based on the Name attribute, or both.
### References

I could not find any, but the original data source was added by this PR
https://github.com/terraform-providers/terraform-provider-aws/pull/8396

### PR DESCRIPTION
**Pull request by DrFaust92**
### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

Closes #11168, Relates #11994

Release note for [CHANGELOG](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md):

```release-note
data_source_aws_vpc_endpoint_service - add filter support
```

Output from acceptance testing:

```
$ make testacc TESTARGS='-run=TestAccDataSourceAwsVpcEndpointService_'
--- PASS: TestAccDataSourceAwsVpcEndpointService_gateway (44.38s)
--- PASS: TestAccDataSourceAwsVpcEndpointService_interface (40.23s)
--- PASS: TestAccDataSourceAwsVpcEndpointService_custom (332.81s)
--- PASS: TestAccDataSourceAwsVpcEndpointService_custom_filter_tags (302.40s)
--- PASS: TestAccDataSourceAwsVpcEndpointService_custom_filter (298.24s)
```

### CODE REVIEW

File: aws/data_source_aws_vpc_endpoint_service.go [hunk 1]
```diff
@@ -69,25 +69,40 @@ func dataSourceAwsVpcEndpointService() *schema.Resource {
...
+		serviceNameOk = true
 	} else if v, ok := d.GetOk("service"); ok {
 		serviceName = fmt.Sprintf("com.amazonaws.%s.%s", meta.(*AWSClient).region, v.(string))
-	} else {
-		return fmt.Errorf(
-			"One of ['service', 'service_name'] must be set to query VPC Endpoint Services")
+		serviceNameOk = true
+	}
+
+	if !filtersOk && !serviceNameOk && !tagsOk {
```
Comment (ewbankkit) [inline aws/data_source_aws_vpc_endpoint_service.go]: Can this now be done at `terraform plan` time with [`Schema.AtLeastOneOf`](https://github.com/hashicorp/terraform-plugin-sdk/blob/ac2f8d01136a1a0198e27d49ea0ea6ed986ea4ee/helper/schema/schema.go#L215-L228)?

**Comment by ewbankkit** [comment]
Verified acceptance tests:

```console
$ make testacc TEST=./aws/ TESTARGS='-run=TestAccDataSourceAwsVpcEndpointService_'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws/ -v -count 1 -parallel 20 -run=TestAccDataSourceAwsVpcEndpointService_ -timeout 120m
=== RUN   TestAccDataSourceAwsVpcEndpointService_gateway
=== PAUSE TestAccDataSourceAwsVpcEndpointService_gateway
=== RUN   TestAccDataSourceAwsVpcEndpointService_interface
=== PAUSE TestAccDataSourceAwsVpcEndpointService_interface
=== RUN   TestAccDataSourceAwsVpcEndpointService_custom
=== PAUSE TestAccDataSourceAwsVpcEndpointService_custom
=== RUN   TestAccDataSourceAwsVpcEndpointService_custom_filter
=== PAUSE TestAccDataSourceAwsVpcEndpointService_custom_filter
=== RUN   TestAccDataSourceAwsVpcEndpointService_custom_filter_tags
=== PAUSE TestAccDataSourceAwsVpcEndpointService_custom_filter_tags
=== CONT  TestAccDataSourceAwsVpcEndpointService_gateway
=== CONT  TestAccDataSourceAwsVpcEndpointService_custom_filter
=== CONT  TestAccDataSourceAwsVpcEndpointService_custom_filter_tags
=== CONT  TestAccDataSourceAwsVpcEndpointService_custom
=== CONT  TestAccDataSourceAwsVpcEndpointService_interface
--- PASS: TestAccDataSourceAwsVpcEndpointService_gateway (23.89s)
--- PASS: TestAccDataSourceAwsVpcEndpointService_interface (24.07s)
--- PASS: TestAccDataSourceAwsVpcEndpointService_custom_filter_tags (254.59s)
--- PASS: TestAccDataSourceAwsVpcEndpointService_custom_filter (265.10s)
--- PASS: TestAccDataSourceAwsVpcEndpointService_custom (265.12s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	265.158s
```

File: aws/data_source_aws_vpc_endpoint_service.go (hunk 1 above)
Comment (DrFaust92) [reply aws/data_source_aws_vpc_endpoint_service.go]: let me check this

Comment (ewbankkit) [reply aws/data_source_aws_vpc_endpoint_service.go]: Or maybe we shouldn't have this (and other ones in other data sources) at all?
It seems many other EC2 data sources don't have such checks and leave it to the user to ensure that enough parameters are supplied to ensure that just 1 resource matches.

Comment (DrFaust92) [reply aws/data_source_aws_vpc_endpoint_service.go]: i took this from data_aws_instance. this is the closest to plan validation as `tags` and `filters` are functions.

Comment (DrFaust92) [reply aws/data_source_aws_vpc_endpoint_service.go]: I thought about it, removing it seems fine. will the aws provided error be descript enough in your opinion?

Comment (ewbankkit) [reply aws/data_source_aws_vpc_endpoint_service.go]: we shouldn't hit an AWS error I don' think.
If there are more than 1 endpoint services that match the parameters (if any) then we will hit the `return fmt.Errorf("multiple VPC Endpoint Services matched; use additional constraints to reduce matches to a single VPC Endpoint Service")`.

Comment (DrFaust92) [reply aws/data_source_aws_vpc_endpoint_service.go]: gotcha, i'll remove it from both PRs.

Comment (DrFaust92) [reply aws/data_source_aws_vpc_endpoint_service.go]: removed & tested.

**Comment by ewbankkit** [comment]
Re-validated after latest changes:

```console
$ AWS_DEFAULT_REGION=us-east-2 make testacc TEST=./aws/ TESTARGS='-run=TestAccDataSourceAwsVpcEndpointService_'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws/ -v -count 1 -parallel 20 -run=TestAccDataSourceAwsVpcEndpointService_ -timeout 120m
=== RUN   TestAccDataSourceAwsVpcEndpointService_gateway
=== PAUSE TestAccDataSourceAwsVpcEndpointService_gateway
=== RUN   TestAccDataSourceAwsVpcEndpointService_interface
=== PAUSE TestAccDataSourceAwsVpcEndpointService_interface
=== RUN   TestAccDataSourceAwsVpcEndpointService_custom
=== PAUSE TestAccDataSourceAwsVpcEndpointService_custom
=== RUN   TestAccDataSourceAwsVpcEndpointService_custom_filter
=== PAUSE TestAccDataSourceAwsVpcEndpointService_custom_filter
=== RUN   TestAccDataSourceAwsVpcEndpointService_custom_filter_tags
=== PAUSE TestAccDataSourceAwsVpcEndpointService_custom_filter_tags
=== CONT  TestAccDataSourceAwsVpcEndpointService_interface
=== CONT  TestAccDataSourceAwsVpcEndpointService_custom_filter_tags
=== CONT  TestAccDataSourceAwsVpcEndpointService_custom_filter
=== CONT  TestAccDataSourceAwsVpcEndpointService_custom
=== CONT  TestAccDataSourceAwsVpcEndpointService_gateway
--- PASS: TestAccDataSourceAwsVpcEndpointService_interface (20.35s)
--- PASS: TestAccDataSourceAwsVpcEndpointService_gateway (21.08s)
--- PASS: TestAccDataSourceAwsVpcEndpointService_custom_filter_tags (229.07s)
--- PASS: TestAccDataSourceAwsVpcEndpointService_custom_filter (238.89s)
--- PASS: TestAccDataSourceAwsVpcEndpointService_custom (249.50s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	249.551s
```

File: website/docs/d/vpc_endpoint_service.html.markdown [hunk 2]
```diff
@@ -41,15 +41,29 @@ data "aws_vpc_endpoint_service" "custome" {
...
 ## Argument Reference
 
 The arguments of this data source act as filters for querying the available VPC endpoint services.
 The given filters must match exactly one VPC endpoint service whose data will be exported as attributes.
 
 * `service` - (Optional) The common name of an AWS service (e.g. `s3`).
 * `service_name` - (Optional) The service name that is specified when creating a VPC endpoint. For AWS services the service name is usually in the form `com.amazonaws.<region>.<service>` (the SageMaker Notebook service is an exception to this rule, the service name is in the form `aws.sagemaker.<region>.notebook`).
+* `filter` - (Optional) One or more name/value pairs to use as filters. There are
+several valid keys, for a full reference, check out
+[describe-vpc-endpoint-services in the AWS CLI reference][1].
```
Comment (bflad) [inline website/docs/d/vpc_endpoint_service.html.markdown]: Can you please update this documentation to the longer form configuration block style? e.g. https://github.com/terraform-providers/terraform-provider-aws/pull/12400/files#diff-259bbf21af9ff942736c706b725f4d6dR76-R88 Thank you!

**Pull request merged**